asmcomp/polling.ml

```diff
@@ -132,47 +168,31 @@ let rec instr_body handlers i =
       desc = Iswitch (index, Array.map instr cases);
       next = instr i.next;
     }
-  | Icatch (Nonrecursive, hdl, body) ->
+  | Icatch (rc, hdl, body) ->
```

Comment:
Because you ignore the rec_flag, you will insert too many polls. For example, in nonrecursive catch handlers that are followed by a nonallocating loop:
```ocaml
let f x y z t =
  begin match x with
  | 0 when y -> z
  | _ -> t
  end;
  while true do ()
  done
```
Here the catch handler of the match gets an unnecessary poll (after fixing the bug).

Reply:
You're right, the rec_flag should be taken into account, and polls should only go on "true" back-edges, not all jumps to handlers.